src/rabbit_vhost.erl

```diff
@@ -213,92 +229,95 @@ assert_benign({error, {absent, Q, _}}, ActingUser) ->
     QName = amqqueue:get_name(Q),
     rabbit_amqqueue:internal_delete(QName, ActingUser).
 
-internal_delete(VHostPath, ActingUser) ->
+internal_delete(VHost, ActingUser) ->
     [ok = rabbit_auth_backend_internal:clear_permissions(
-            proplists:get_value(user, Info), VHostPath, ActingUser)
-     || Info <- rabbit_auth_backend_internal:list_vhost_permissions(VHostPath)],
-    TopicPermissions = rabbit_auth_backend_internal:list_vhost_topic_permissions(VHostPath),
+            proplists:get_value(user, Info), VHost, ActingUser)
+     || Info <- rabbit_auth_backend_internal:list_vhost_permissions(VHost)],
+    TopicPermissions = rabbit_auth_backend_internal:list_vhost_topic_permissions(VHost),
     [ok = rabbit_auth_backend_internal:clear_topic_permissions(
-        proplists:get_value(user, TopicPermission), VHostPath, ActingUser)
+        proplists:get_value(user, TopicPermission), VHost, ActingUser)
      || TopicPermission <- TopicPermissions],
-    Fs1 = [rabbit_runtime_parameters:clear(VHostPath,
+    Fs1 = [rabbit_runtime_parameters:clear(VHost,
                                            proplists:get_value(component, Info),
                                            proplists:get_value(name, Info),
                                            ActingUser)
-     || Info <- rabbit_runtime_parameters:list(VHostPath)],
-    Fs2 = [rabbit_policy:delete(VHostPath, proplists:get_value(name, Info), ActingUser)
-           || Info <- rabbit_policy:list(VHostPath)],
-    ok = mnesia:delete({rabbit_vhost, VHostPath}),
+     || Info <- rabbit_runtime_parameters:list(VHost)],
+    Fs2 = [rabbit_policy:delete(VHost, proplists:get_value(name, Info), ActingUser)
+           || Info <- rabbit_policy:list(VHost)],
+    ok = mnesia:delete({rabbit_vhost, VHost}),
     Fs1 ++ Fs2.
 
--spec exists(rabbit_types:vhost()) -> boolean().
+-spec exists(vhost:name()) -> boolean().
+
+exists(VHost) ->
+    mnesia:dirty_read({rabbit_vhost, VHost}) /= [].
 
-exists(VHostPath) ->
-    mnesia:dirty_read({rabbit_vhost, VHostPath}) /= [].
+-spec list_names() -> [vhost:name()].
 
--spec list() -> [rabbit_types:vhost()].
+list_names() -> mnesia:dirty_all_keys(rabbit_vhost).
 
-list() ->
-    mnesia:dirty_all_keys(rabbit_vhost).
+-spec list() -> [vhost:name()].
+
+list() -> mnesia:dirty_match_object(rabbit_vhost, vhost:pattern_match_all()).
```

Comment:
Can we keep the `list()` function unchanged, and add another function to list records? For the sake of keeping the API stable between 3.7.x and 3.8.x.

Reply:
Done.